text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,407 @@
+```
+[[processes]]
+type = "web"
+command = "dotnet my-app.dll --urls http://0.0.0.0:${PORT:-8080}"
+direct = false
+```
+NOTE: the buildpack API used by this buildpack (`0.5`) predates the introduction of `default`.
+
+Using the new API this process could look like:
+```
+[[processes]]
+type = "web"
+command = ["dotnet", "my-app.dll", "--urls", "http://0.0.0.0:${PORT}"] # the default value of PORT would need to be provided in a layer
+default = true
+```
+Things to note:
+* In the above example I have eliminated the dependency on Bash instead of explicitly adding it to the command, because it is likely unnecessary.
+* If the buildpack authors believed that `--urls` should be overridable they could set move the last two arguments from `command` to `args`.
+
+### Example 2 - A Script Process
+When a buildpack upgrade to the new buildpack API, it must convert any existing shell processes into direct processes.
+
+The [Paketo Yarn Start Buildpack](https://github.com/paketo-buildpacks/yarn-start) currently may generate a script processes similar to the following:
+
+```
+[[processes]]
+type = "web"
+command = "pre-start.sh && nodejs server.js && post-start.sh"
+direct = false
+default = false
+```
+
+Using the new API this process look like:
+```
+[[processes]]
+type = "web"
+command = ["bash", "-c", "pre-start.sh && nodejs server.js && post-start.sh"]
+default = false
+```
+
+## User Provided Processes
+
+Currently if the user can specify a custom process dynamically at runtime by setting the container entrypoint to `launcher` directly rather than using a symlink to the launcher, the providing a custom `cmd`. This custom command is executed directly if `cmd` is an array and the first element is `--`. Otherwise the custom command is assumed to be a shell process. In the interest of removing complexity we should do away with the special `--` argument and execute all custom commands directly. 
+
+### Example 1 - A Direct process
+The follow direct commands:
+```
+docker run --entrypoint launcher <image> -- env
+docker run --entrypoint launcher <image> -- echo hello '$WORLD' 
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> env
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+```
+Previously, in the second command in this example `${WORLD}` would not have been interpolated because this is a direct process; instead the output would include the literal string `$WORLD`. With the changes proposed, `${WORLD}` will now be evaluated, even though the process is direct.
+
+### Example 2 - A Shell Process
+The follow custom shell command:
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> echo hello '${WORLD:-world}'
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> bash -c 'echo hello "${WORLD:-world}"'
```

Comment:
Could we support these use cases by default? GIven that most envsubst libs already support it - 

https://github.com/a8m/envsubst
https://github.com/drone/envsubst

This would also be useful as it would end up in the metadata and users could see default values for these variables as set by processes.

Reply:
I _just_ decided to take this in a different direction - https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3

Reply:
I found some of the comment from the original k8s implementation of this illuminating - https://github.com/kubernetes/kubernetes/pull/7936#issuecomment-100950516

Reply:
Was going through the same thing.


On a separate note - I guess it should be a trivial change in the above libs to add support for a different marker instead of `{`. WIll see if they are open to it. OTOH I guess we could always start simple and add it later if needed. I would've like a way to specify a default value at the very least though.

Reply:
FWIW - just needs a change at https://github.com/drone/envsubst/blob/7bf45dbf5372da60a12eab36b9c8a63c99b92ea5/parse/scan.go#L174

Reply:
@samj1912 can this comment be resolved? or do you still want to advocate for more env substitution functionality, including defaults?

Personally I think we should go with the simple version for now.